rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,265 @@
+{ "host": "localhost", "message": "bar" }
+```
+
+This would also be available as an `unnest` transform for simple cases where the user does not need to do any additional mapping:
+
+```toml
+[transforms.unnest]
+type = "unnest"
+field = "events"
+```
+
+Which would have the same effect. I think the implementation could just take advantage of the "expand" feature of transforms that lets them replace themselves at vector start time with another transform.
+
+Additionally, we will provide `only_fields` and `except_fields` as options on the `unnest` function and transform to allow users to select which fields will be kept. These match similar semantics to the `encoding` options on sinks.
+
+Example input:
+
+```json
+{ "timestamp": "2020-12-09T16:09:53+00:00", "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Remap transform config:
+
+```toml
+[transforms.remap]
+type = "remap"
+source = """
+. = unnest(., "events", only_fields: ["host"])
+"""
+```
+
+Ouput:
+
+```json
+{ "host": "localhost", "message": "foo" }
+{ "host": "localhost", "message": "bar" }
+```
+
+Here the `timestamp` field is not preserved.
+
+## Doc-level Proposal
+
+To convert an incoming event into multiple events, the `unnest` transform can be used.
+
+For example, given an input of:
+
+```json
+{ "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }, 1] }
+```
+
+And a transform of:
+
+```toml
+[transforms.unnest]
+type = "unnest"
+field = "events"
+```
+
+The following events will be output:
+
+```json
+{ "host": "localhost", "message": "foo" }
+{ "host": "localhost", "message": "bar" }
+{ "host": "localhost", "message": "1" }
+```
+
+That is, each record in the indicated field will be emitted as its own event, merged with any other fields existing at the top-level of the event.
+
+If any elements in the array field are not an object, they will be turned into a string and set as the `message` key.
+
+The `remap` transform can also be used to emit multiple events from a single incoming event by setting the root path, `.`, to an array. The above transform is synonymous with:
+
+```toml
+[transforms.remap]
+type = "remap"
+source = """
+. = unnest(., "events")
+"""
+```
+
+## Rationale
+
+This enhances `remap` to be able to emit multiple events as well as provides a `unnest` transform for simple cases. Without this, users will continue to have to use Lua or WASM to acheive this, which introduces a performance bottleneck compared to this proposal.
+
+## Prior Art
+
+- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
+- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)
+
+These are similar to the proposed approach.
+
+## Drawbacks
+
+- Adds an additional transform to be aware of.
+- Less flexible than `emit_log` alternative which could emit arbitrary events from `remap` transform.
+- Ongoing maintenance burden should be minimal.
```

Comment:
An additional drawback of this solution as opposed to the previous one (separate transform, using final value of program), is that it _can_ result in unexpected outcomes if someone is doing something like this:

```jsonl
// most common events
{ "message": "{ \"foo\": true }" }

// uncommon, unexpected
{ "message": "[true, false]" }
```

```swift
. = parse_json!(.message)
```

output:

```jsonl
// expected
{ "foo": true }
```

```jsonl
// unexpected
{ "message": true }
{ "message": false }
```

By having exploding/unnesting be done in a dedicated transform, people explicitly opt-in to this behaviour, which can still cause unexpected results for badly formatted events, but it means (1) that when you see multiple events you know exactly where to debug, and (2) when using the `remap` transform, you never have to worry about the above situation occurring.

This is why I'm still in favour of the previous proposal from a useability perspective, but I'm not _opposed_ to the current approach, so I'm happy to see this land if others prefer this over the alternative.

Reply:
I should note that the above example is something I've seen happen plenty of times as I worked on similar observability systems in the past (e.g. tens/hundreds of services, managed by different teams, using their own techniques/libraries to emit events, some untyped, and accidentally emitting non-conforming event shapes).

Reply:
🤔 That is a good observation.

I think generally people neglect that `parse_json` can return other things than just objects. I think maybe we could improve the docs for that function as a start to have examples of enforcing the expected type through the type assertion functions.

I agree a separate transform would make it more obvious, but I also like the idea of users just being able to use remap.

Reply:
One approach to address this could be adding an explicit flag (e.g. `emit_multiple`) to the transform config, which would error when anything other than an array of objects is present in the VRL root after transformation.

Or vice versa error when an array is present but the flag was not set.